### DISCUSSION TITLE
Makros in Kopf- und Fußtext

### DISCUSSION THREAD

**Discussion by dragotin** [dragotin/kraft#208] [Show and tell]
Mit einem der letzten Commits wurde Kraft um eine mächtige neue Funktion erweitert:  In den Einleitungs- und Schlusstexten der Dokumente von Kraft werden jetzt Makros unterstützt, mit denen sich interessante neue Funktionalitäten umsetzen lassen.

Makros sind einfach "Anweisungen", die durch Kraft unter Berücksichtigung des Dokumentes ersetzt werden. Ein Beispiel ist ein Zahlungsziel, das 14 Tage nach dem Datum der Rechnung liegen soll. Bisher musste das manuell ermittelt und in den Schlußtext der Rechnung eingetragen werden. Jetzt kann das per Makro in der Text-Vorlage automatisiert werden.

## Datumsrechnung

Das oben beschriebene wird mit dem Macro `DATE_ADD_DAYS(<anzahl_tage>)` umgesetzt. Im Abschlußtext einer Rechnung könnte also stehen:

>> "Bitte überweisen Sie den Rechnungsbetrag bis DATE_ADD_DAYS(14) auf unser Konto."

Das Makro wird durch das Datum, das 14 Tage nach dem Rechnungsdatum liegt, ersetzt.

## Haushaltsnahe Dienstleistungen

Die Summe bestimmter Posten einer Rechnung sollen auf der Rechnung gesondert ausgewiesen werden, zum Beispiel für die Arbeitszeit-Kosten im Rahmen der sog. Haushaltsnahen Dienstleistungen.

Dabei hilft eine weitere Gruppe von Macros.

Voraussetzung ist, dass die Posten der Rechnung, die zu den Arbeitszeitkosten gezählt werden sollen, ensprechend markiert werden. Dazu werden die Posten in Kraft mit dem Tag _"Work"_ versehen.

![grafik](https://github.com/dragotin/kraft/assets/1070214/90d258d8-3c3d-4df0-9397-d2b0a235b62f)

Dann kann dem Fußtext der Rechnung folgendes Macro hinzugefügt werden:

>> Diese Rechnung enthält einen Anteil von SUM_PER_TAG(Work) an Arbeitszeit-Kosten.

Das Makro `SUM_PER_TAG(<tagname>)` summiert die Netto-Summe aller Posten, die mit der entsprechenden Markierung versehen sind, und gibt diese aus.

Um hinzuzufügen, aus wie vielen Posten sich die Summe zusammensetzt, gibt es das Makro `ITEM_COUNT_WITH_TAG(<tagname>)`.

Um keinen zusätzlichen Text anzuzeigen, wenn es keine Work-Posten gibt, kann das Makro `IF_ANY_HAS_TAG(<tagname>)` eingesetzt werden. Der Text zwischen diesem Makro und seinem schliessenden Teil `END_HAS_TAG` wird nur angezeigt, wenn es überhaupt einen Posten in dem Dokument gibt, der mit dem tagnamen markiert wurde.

Zusammenfassend kann eine Fußtext-Vorlage dann so erweitert werden:

>> IF_ANY_HAS_TAG(Work)
>> Diese Rechnung enthält ITEM_COUNT_WITH_TAG(Work) Arbeitszeit-Posten mit einer Gesamtsumme von SUM_PER_TAG(Work).
>> END_HAS_TAG

Bestimmt können mit dieser Erweiterung noch andere nützliche Funktionen umgesetzt werden.

Ich freue mich über Feedback zu dem Themenbereich.

**Comment by noseshimself** [comment]
Kann man Macros selbst definieren? Beispielsweise "Rechnung" in "das Dokument einer Art, die nicht genannt werden darf, ohne daß Geld fließen muß"?

**Reply by dragotin** [reply]
Hm, ich versteh nicht wirklich, was du meinst. Aber nein, bisher sind die Makros nicht frei definierbar, sondern hart codiert.

**Comment by dragotin** [comment]
Kleine Anpassung:

Bisher gab es nur das Makro `SUM_PER_TAG(<Tag>)`, das die Netto Summe der Posten mit dem Tag ausgegeben hat. Bei den haushaltsnahen Dienstleistungen ist es aber üblich, auch den Brutto-Betrag mit anzugeben. Deshalb gibt es jetzt drei Makros dafür:
- `NETTO_SUM_PER_TAG(<Tag>)` - Netto Summe aller Posten, die mit dem Tag markiert sind
- `BRUTTO_SUM_PER_TAG(<Tag>)` - Brutto Summe
- `VAT_SUM_PER_TAG(<Tag>)` - Nur der Steueranteil der Posten mit dem Tag.

Entsprechend würde dann das Snippet für die HnD so aussehen:
```
IF_ANY_HAS_TAG(HnD)
Diese Rechnung enthält ITEM_COUNT_WITH_TAG(HnD) Arbeitszeit-Posten mit einer Netto-Summe von 
NETTO_SUM_PER_TAG(HnD) plus VAT_SUM_PER_TAG(HnD) Umsatzsteuer ergibt den Brutto Betrag von
 BRUTTO_SUM_PER_TAG(HnD).
END_HAS_TAG
```
ACHTUNG: Das Makro `NETTO_SUM_PER_TAG()` muss enthalten sein, damit die anderen funktionieren!

**Reply by Reiner-W123** [reply]
Lieben dank für die Makros.
Jetzt brauche ich eine weiter Entwicklung.  Bis jetzt werden nur die Stundenpositionen  für die HnD zusammen gerechnet. Aber wie macht man es mit Kalkulieren Positionen? Bsp. 4€ Lohn auf m² + 1€ Material auf m² = 5€ m². Hier muss das Material für die HnD herausgezogen werden.

**Reply by dragotin** [reply]
Ja, die Arbeitszeit aus den Kalkulationsdaten der in den Dokumenten verwendeten Vorlagen zu ziehen ist noch nicht möglich. Das ist auch nicht einfach, da sich Kraft dafür für jeden Posten die richtige "Version" der kalkulierten Vorlage merken müsste.

In Kraft 2.0 wird es ein bisschen ein geändertes Daten-Modell geben, mit dem wird das einfacher zu lösen sein. Das wird aber noch etwas dauern.

**Comment by bastelinsky** [comment]
Muss man irgendetwas aktivieren, damit die Makros berücksichtigt werden?
Bei mir funktioniert das nicht:
"Bitte überweisen Sie den Rechnungsbetrag bis zum DATE_ADD_DAYS(14) auf unser Konto:"
(Linuxmint 21.2, Kraft-Appimage 1.0)

**Reply by noseshimself** [reply]
Und was steht im Template?

**Reply by dragotin** [reply]
Du brauchst Version 1.1, in 1.0 ist das noch nicht drin. 

Versuch mal das AppImage von https://github.com/dragotin/kraft/releases/tag/v1.1

**Reply by bastelinsky** [reply]
Danke für Eure Unterstützung. 
Mit Version 1.1 funktioniert es bestens.
